stackslib/src/chainstate/burn/db/sortdb.rs

```diff
@@ -3468,6 +3468,96 @@ impl SortitionDB {
         Ok(())
     }
 
+    pub fn find_first_prepare_phase_sortition(
+        &self,
+        from_tip: &SortitionId,
+    ) -> Result<Option<SortitionId>, db_error> {
+        let from_tip =
+            SortitionDB::get_block_snapshot(self.conn(), &from_tip)?.ok_or_else(|| {
+                error!(
+                    "Could not find snapshot for sortition";
+                    "sortition_id" => %from_tip,
+                );
+                db_error::NotFoundError
+            })?;
+        let mut cursor = from_tip;
+        let mut last = None;
+        while self
```

Comment:
I'm not sure why you need this loop here?   Given a sortition at a given height, you can calculate what reward cycle it's in, and from there work out the height of the first prepare phase block within that reward cycle (`SortitionDB` (`self`) has a `PoxConstants` instance and a `first_burn_height` value so you can easily compute it).